addons/Dexie.Observable/src/Dexie.Observable.js

```diff
@@ -254,19 +254,23 @@ function Observable(db) {
                             // Take over mastership
                             mySyncNode.node.isMaster = 1;
                             currentMaster.isMaster = 0;
-                            return db._syncNodes.put(currentMaster);
+                            db.transaction('rw!', '_syncNodes', () => {
+                                return db._syncNodes.put(currentMaster);
+                            });
                         }
                     }).then(()=>{
-                        // Add our node to DB and start subscribing to events
-                        return db._syncNodes.add(mySyncNode.node).then(function() {
-                            Observable.on('latestRevisionIncremented', onLatestRevisionIncremented); // Wakeup when a new revision is available.
-                            Observable.on('beforeunload', onBeforeUnload);
-                            Observable.on('suicideNurseCall', onSuicide);
-                            Observable.on('intercomm', onIntercomm);
-                            // Start polling for changes and do cleanups:
-                            pollHandle = setTimeout(poll, LOCAL_POLL);
-                            // Start heartbeat
-                            heartbeatHandle = setTimeout(heartbeat, HEARTBEAT_INTERVAL);
+                        db.transaction('rw!', '_syncNodes', () => {
```

Comment:
This 'works', but needs a bit of scrutiny as to whether it retains atomicity:

- https://dexie.org/docs/Dexie/Dexie.transaction()#specify-reusage-of-parent-transaction
- https://dexie.org/docs/Dexie/Dexie.transaction()#implementation-details-of-nested-transactions

Ideally some test coverage to check & prove atomicity would be nice.

Reply:
Another related thought: this is currently essentially performing a get-and-set within a transaction to ensure atomicity.

Would it be possible instead to use a single write operation ([`Table.modify`](https://dexie.org/docs/Collection/Collection.modify()) or similar) to perform the required updates?

That way there would be no read-only query required within the transaction, and the transaction would not risk becoming inactive as a result of that query completing before any writes occur.

Our requirements are to:

- Set the current node's `isMaster` flag when no record with `isMaster` and an up-to-date heartbeat timestamp is present
- Unset the `isMaster` flag on a record if the heartbeat timestamp is out-of-date
- Write the current node record into the table

The challenge is that the value we write for the current node record depends on the contents of a potentially existing (or potentially nonexistent) record with `isMaster` set and potentially out-of-date heartbeat timestamp.

Reply:
It does look like the `isMaster` field is an [indexable type](https://dexie.org/docs/Indexable-Type) (ref: [definition as `Number`](https://github.com/dfahlander/Dexie.js/blob/77f0b08c58784bfaaf9e2f5e26dd4e1d7b7d3094/addons/Dexie.Observable/src/Dexie.Observable.d.ts#L96)), so perhaps it would be possible to use [`db._syncNodes.orderBy(...).modify(...)`](https://dexie.org/docs/Table/Table.orderBy()) to retrieve existing syncnodes in order, maintain an external variable to store the 'current node should become master' state, and then use a function argument to [modify](https://dexie.org/docs/Collection/Collection.modify()) to write the expected object record(s).